### PR TITLE
Feature/authentication

### PR DESCRIPTION
**Pull request by kirill-oleynik**

### CODE REVIEW

File: config/initializers/jwt_sessions.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+JWTSessions.token_store = :redis, {
+  redis_host: ENV['REDIS_HOST'],
+  redis_port: ENV['REDIS_PORT'],
+  redis_db_name: ENV['REDIS_DB_NAME'],
+  token_prefix: ENV['TOKEN_PERFIX']
+}
+JWTSessions.algorithm = 'HS256'
+JWTSessions.encryption_key = ENV['JWTSESSIONS_ENCRYPTION_KEY']
+JWTSessions.access_exp_time = EXPIRATION[:access]
+JWTSessions.refresh_exp_time = EXPIRATION[:refresh]
```
Comment (StanisLoveSid) [inline config/initializers/jwt_sessions.rb:16]:
```ruby
JWTSessions.access_exp_time = 1.hour
JWTSessions.refresh_exp_time = 24.hours
```

File: config/initializers/jwt_sessions.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+EXPIRATION = {
+  access: 3_600, # 1 hour
+  refresh: 86_400 # 24 hours
+}.freeze
+JWTSessions.token_store = :redis, {
+  redis_host: ENV['REDIS_HOST'],
+  redis_port: ENV['REDIS_PORT'],
+  redis_db_name: ENV['REDIS_DB_NAME'],
+  token_prefix: ENV['TOKEN_PERFIX']
```
Comment (StanisLoveSid) [inline config/initializers/jwt_sessions.rb:11]: let's use rails credentials for env cars keeping

File: app/services/jwt_session_builder.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+# Responsible for building jwt user session
+class JwtSessionBuilder
```
Comment (StanisLoveSid) [inline app/services/jwt_session_builder.rb:4]: keep in mind extending jwt session logic

**Review by StanisLoveSid — APPROVED** [review]

File: spec/factories/users.rb [hunk 4]
```diff
@@ -5,6 +5,6 @@
     f.username { Faker::Name.unique.name }
     f.email { Faker::Internet.unique.email }
     f.password { Faker::Internet.password }
-    f.password_confirmation { password }
+    # f.password_confirmation { password }
```
Comment (kirill-oleynik) [inline spec/factories/users.rb:8]: @StanisLoveSid why did you do that? That That is the reason of failing tests.
In the future, be sure to run tests locally and make sure everything works before pushing in your commits!

File: app/services/user_input_error.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+          if value[0].is_a?(Array)
+            add_error.call([value[0][0], value[0][1]], {}).deep_transform_keys { |k| k.to_s.camelize(:lower) }
+          else
+            value
+          end
+        errors[key.to_s.camelize(:lower)] = error_message
+        errors
+      end
+    end
+  end
```
Comment (kirill-oleynik) [inline app/services/user_input_error.rb:40]: @StanisLoveSid what is the purpose of this class? why do we need it?
We have GraphQL::UserInputError.

**Pull request merged**